### ISSUE TITLE
Couple of benchmarks 

### ISSUE THREAD

**Issue by Albul** [brianburton/java-immutable-collections#19]
Here I have create couple of benchmarks, it includes standard jdk as well as pcollections and   ##java-immutable-collections:
https://github.com/Albul/benchmarks-java-kotlin

**Comment by brianburton** [comment]
Very interesting Albul!  Thanks so much for posting this!  Looks like the library did well overall but there are areas to work on.  I'd like to improve the map iteration speed for example.

**Comment by brianburton** [comment]
Thanks for the size measurements.  Nice to see how little space overhead there is to JImmutableTreeList and JImmutableTreeMap.  The JImmutableHashMap size is expected since the collision maps it uses internally add some extra space overhead.  I've considered reducing that but for now it seems acceptable.  It's really not that much larger than the guava map which is remarkable given the latter's inflexibility.

Set maps and list maps are maps of sets/lists so naturally they are quite a bit larger than single value maps.  Maybe they could go into a separate category along with multi-maps and multi-sets from the various libraries?

**Comment by brianburton** [comment]
@Albul I've been using my social distancing time off to dramatically reduce memory consumption by hash maps and sets.  Your benchmarks have been very helpful for me in finding the hotspots of memory consumption.  I was able to reduce memory consumption by something like 34% for those collections.

Based on my tests the changes have put JImmutableHashMap and JImmutableHashSet only very slightly higher memory usage than the standard JDK HashMap and HashSet.  This is a snippet from a test using the current development build.  That's only around a 3.2% memory difference between the immutable and mutable classes as compared to about 58% difference in the 3.0.2 release.

````
Object type: HashMap 5846608
Object type: HashSet 5846624
Object type: JImmutableHashMap 6032632
Object type: JImmutableHashSet 6032632
````
I'm going to let the stress tests run for a day or two before releasing the changes.

**Comment by brianburton** [comment]
@Albul I have released 3.1.0 now if you want to update your benchmarks.

**Comment by brianburton** [comment]
The 3.2.0 release has further reduced the memory footprint of hash maps and sets.  Both now have smaller memory footprints than their standard java mutable counterparts!

````
Object type: HashMap 5846608
Object type: HashSet 5846624
Object type: JImmutableHashMap 4512936
Object type: JImmutableHashSet 3712936
````

Interestingly the JImmutableTreeList uses only slightly larger (approx. 4%) more memory than a mutable ArrayList.

````
Object type: ArrayList 1997992
Object type: JImmutableTreeList 2064624
````
